### PR TITLE
inspection frames

### PR DESCRIPTION
**Pull request by loechel**

### CODE REVIEW

**Comment by loechel** [comment]
Feedback welcome.

* potential more documentation necessary on what to check on new Python Versions
* Changelog Entry
* maybe move / rename the new test file

File: docs/contributing/index.rst [hunk 1]
```diff
@@ -64,6 +64,7 @@ To do so:
     All AST Nodes without an explicit ``visit_<AST Node>`` method, are denied by default.
     So the usage of this expression and functionality is not allowed.
 
+* check the documentation for ``inspect`` <https://docs.python.org/3/library/inspect.html> and adjust the ``transformer.py:INSPECT_ATTRIBUTES`` list
```
Comment (icemac) [inline docs/contributing/index.rst:67]:
```suggestion
* Check the documentation for `inspect  <https://docs.python.org/3/library/inspect.html>`_ and adjust the ``transformer.py:INSPECT_ATTRIBUTES`` list.
```

File: src/RestrictedPython/transformer.py [hunk 2]
```diff
@@ -63,6 +63,35 @@
     'breakpoint',
 ])
 
+INSPECT_ATTRIBUTES = frozenset([
+    # inspect attributes:
+    # traceback
+    "tb_frame",
+    "tb_lasti",
+    "tb_lineno",
```
Comment (icemac) [inline src/RestrictedPython/transformer.py:71]:
```suggestion
    # "tb_lasti",  # just an int
    # "tb_lineno",  # just an int
```

File: src/RestrictedPython/transformer.py [hunk 3]
```diff
@@ -63,6 +63,35 @@
...
+    "tb_next",
+    # frame
+    "f_back",
+    "f_builtins",
+    "f_code",
+    "f_globals",
+    "f_lasti",
+    "f_lineno",
```
Comment (icemac) [inline src/RestrictedPython/transformer.py:79]:
```suggestion
    # "f_lasti",  # just an int
    # "f_lineno",  # just an int
```

File: src/RestrictedPython/transformer.py [hunk 4]
```diff
@@ -63,6 +63,35 @@
...
+    "f_locals",
+    "f_trace",
+    # generator
+    "gi_frame",
+    "gi_running",
```
Comment (icemac) [inline src/RestrictedPython/transformer.py:84]:
```suggestion
    # "gi_running", # just a bool
```

File: src/RestrictedPython/transformer.py [hunk 5]
```diff
@@ -63,6 +63,35 @@
...
+    # frame
+    "f_back",
+    "f_builtins",
+    "f_code",
+    "f_globals",
+    "f_lasti",
+    "f_lineno",
+    "f_locals",
+    "f_trace",
+    # generator
```
Comment (icemac) [inline src/RestrictedPython/transformer.py:82]:
```suggestion
    # code
    "co_code",
    # generator
```

File: src/RestrictedPython/transformer.py [hunk 6]
```diff
@@ -63,6 +63,35 @@
...
+    "gi_frame",
+    "gi_running",
+    "gi_code",
+    "gi_yieldfrom",
+    # coroutine
+    "cr_await",
+    "cr_frame",
+    "cr_running",
```
Comment (icemac) [inline src/RestrictedPython/transformer.py:90]:
```suggestion
    # "cr_running",  # just a bool
```

File: src/RestrictedPython/transformer.py [hunk 7]
```diff
@@ -844,6 +873,15 @@ def visit_Attribute(self, node):
                 '"{name}" is an invalid attribute name because it ends '
                 'with "__roles__".'.format(name=node.attr))
 
+        if node.attr in INSPECT_ATTRIBUTES:
+            self.error(
+                node,
+                '"{name}" is a reserved inspect attribute name, '
+                'that is forbidden to call in RestrictedPython'.format(
+                    name=node.attr,
```
Comment (icemac) [inline src/RestrictedPython/transformer.py:881]:
```suggestion
                f'"{node.attr}" is a restricted name,'
                ' that is forbidden to access in RestrictedPython.'
```

**Review by icemac — CHANGES_REQUESTED** [review]
Thank you for tackling this issue.

I have some suggestions, gathered by trying it out in a Python interpreter.
Changing the error message also requires changing the tests.

**Comment by loechel** [comment]
@icemac, thanks for your comments and improvements recommendations.
I will close this pull request, see mail and follow up there.

**Pull request closed without merging**